File: src/test/resources/bdd/features/steps/steps.py
__author__ = 'jacekf'

from behave import *
from nose.tools import *
from difflib import Differ, unified_diff
import httplib2, json, sys, difflib, urllib, urlparse, string, codecs

from urllib import urlencode
import httplib, base64, string, unicodedata

SKIP = '_'

###################################################################
# GIVEN
###################################################################

@given('"{name}:{password}" sends {method} "{url}"')
def given_i_post_to_url(context, name, password,method, url):
    """
    Used for setting up test data
    """
    for row in context.table:
        args = []
        for key in row.headings:
            val = row[key]
            if unicode(val) is not unicode(SKIP):
                args.append("%s=%s" % (key,val))

        # convert it to an HTML format form, to be compatible with existing methods
        form = "&".join(args)
        when_i_post_put_to_url_with_params(context,name,password,method,url,form)
        # validate was successful
        if str(method) == "POST":
            then_i_expect_http_code(context,201)
        else:
            then_i_expect_http_code(context,200)


###################################################################
# WHEN
###################################################################
@when('"{name}:{password}" sends {method} "{url}" with "{params}"')
def when_i_post_put_to_url_with_params(context, name, password,method, url, params = SKIP, admin = False, content_type = "application/x-www-form-urlencoded"):
    full_url = "%s%s" % (context.app_config.app_url_prefix,url) if admin is False else "%s%s" % (context.app_config.admin_url_prefix,url)
    h = httplib2.Http()
    h.add_credentials(name,password)
    # convert string params to dictionary
    form_params = {}
    if params is not SKIP:
        form_params = dict(urlparse.parse_qsl(params))

    # handle both
    content = params.encode("utf-8")

    if content_type is "application/x-www-form-urlencoded":
        content = urllib.urlencode(form_params)

    if method in {"PUT","POST","PATCH"}:
        context.http_headers["content-type"] = content_type

    # encode http headers
    context.http_headers = dict((k.encode('ascii') if isinstance(k, unicode) else k,
                                 v.encode('ascii') if isinstance(v, unicode) else v)
        for k,v in context.http_headers.items())

    context.resp, context.content = h.request(full_url.encode("utf-8"), method.encode("utf-8"), content, headers=context.http_headers)
    context.http_headers = {}

@when('"{name}:{password}" sends {method} "{url}" with {file_type} file "{file_path}"')
def when_i_send_file(context,name,password,method,url,file_type,file_path):
    content = None

    with codecs.open(file_path, encoding='utf-8') as f:
        content = f.read()

    content_type = "text/plain; charset=utf-8"
    if file_type in ("CSV","csv"):
        content_type = "text/csv; charset=utf-8"

    when_i_post_put_to_url_with_params(context, name, password,method, url, params = content, content_type = content_type)

@when('"{name}:{password}" sends {method} "{url}" with {type}')
def when_i_send_text(context,name,password,method,url,type):
    content = context.text

    content_type = "text/plain; charset=utf-8"
    if type in ("JSON","json"):
        content_type = "application/json; charset=utf-8"

    when_i_post_put_to_url_with_params(context, name, password,method, url, params = content, content_type = content_type)


@when('"{name}:{password}" sends {method} "{url}"')
def when_i_send_to_app_port(context,name,password,method,url):
    when_i_post_put_to_url_with_params(context, name, password,method, url)

@when('"{name}:{password}" sends {method} "{url}" on admin port')
def when_i_send_to_admin_port(context,name,password,method,url):
    when_i_post_put_to_url_with_params(context, name, password,method, url, admin= True)

@when('I set HTTP header "{header_name}" to "{header_value}"')
def when_i_set_http_header(context,header_name,header_value):
    context.http_headers[header_name] = header_value


###################################################################
# THEN
###################################################################
@then('I expect HTTP code {code}')
def then_i_expect_http_code(context,code):
    error_message = u"Expected HTTP code %s != %s\nResponse:\n%s\n%s" % (code, context.resp.status, context.resp, context.content.decode('utf-8'))
    assert_equals(int(code),context.resp.status, error_message)

@then('I expect JSON equivalent to')
def and_i_expect_json_equivalent_to(context):
    """This method looks only at the field names that are specified in the expected JSON. All other field names are removed for comparison purposes.
        This shields the JSON from changes over time, when new fields are being added
    """

    actual_json = json.loads(context.content)
    exp_json = json.loads(context.text)

    # remove fields that are not explicitly listed in the expected JSON
    # to allow APIs to add new fields without breaking existing tests
    # remove the additional fields from the received JSON
    expected_fields = get_unique_fields(exp_json, set())
    received_fields = get_unique_fields(actual_json, set())
    delta_fields = received_fields - expected_fields

    original_json = json.dumps(actual_json, sort_keys=True, indent=4)

    remove_property_list_from_object_recursively(actual_json, delta_fields)

    expected_json_sorted = json.dumps(exp_json, sort_keys=True, indent=4)
    actual_json_sorted = json.dumps(actual_json, sort_keys=True, indent=4)

    # prepare potential error message
    assert_json(expected_json_sorted,actual_json_sorted,original_json)

@then('I expect HTTP header "{header}" equals "{value}"')
def then_i_expect_http_header_value(context,header,value):
    if str(value) != str(SKIP):
        actual_header = context.resp[header]
        assert_equals(value,actual_header)

@then('I expect HTTP content equals "{content}"')
def then_i_expect_http_content_equals(context,content):
    if str(content) is not SKIP:
        assert_equals(content,context.content)

@then('I expect HTTP content contains "{content}"')
def then_i_expect_http_content_contains(context,content):
    error_message = "Content '%s' not found\nResponse:\n%s\n%s" % (content,context.resp,context.content)
    if str(content) is not SKIP:
        assert_true(content in context.content,error_message)

###################################################################
# UTILITY METHODS
###################################################################

#This removes recursively from the object parameter all the properties contained in the propertiesToRemove list
def remove_property_list_from_object_recursively(object, propertiesToRemove):
    if type(object).__name__ == 'list':
        for subObject in object:
            remove_property_list_from_object_recursively(subObject, propertiesToRemove)
    elif type(object).__name__ == 'dict':
    #This is in case a property contains another object
        for key in object.keys():
            remove_property_list_from_object_recursively(object[key], propertiesToRemove)

        for property in propertiesToRemove:
            remove_property_from_object(object, property)


def get_unique_fields(list_or_dict, fieldset):
    "Traverses an entire JSON documents and returns a tuple of all the unique field names, no matter how deep"
    if type(list_or_dict).__name__ == 'list':
        for subObject in list_or_dict:
            get_unique_fields(subObject, fieldset)
    elif type(list_or_dict).__name__ == 'dict':
    #This is in case a property contains another list_or_dict
        for key in list_or_dict.keys():
            fieldset.add(key)
            get_unique_fields(list_or_dict[key], fieldset)
    return fieldset

#This will remove the propertyToRemove property from the list_or_dict parameter
def remove_property_from_object(object, propertyToRemove):
    try:
        if(object[propertyToRemove] != None):
            del object[propertyToRemove]
    except Exception:
        return

def diff(expected, actual):
    expected = expected.splitlines(1)
    actual = actual.splitlines(1)
    diff = unified_diff(expected, actual)
    return ''.join(diff)

def assert_json(expected, actual, full_actual_json=None):
    # allow passing FULL JSON before delta fields were removed
    if full_actual_json == None:
        full_actual_json = actual
    try:
        assert_equals(expected, actual)
    except AssertionError, e:
        raise AssertionError("Expected JSON:\n%s\n*** actual ****\n%s\n*** diff on expected fields only ****\n%s" % (
            expected, full_actual_json, diff(expected, actual)))

File: src/test/resources/bdd/features/steps/steps.rb
require "net/http"
require "json"

SKIP = "_"

###### GIVEN #############

Given(/^"(.+):(.+)" sends POST "([^"]+)":/) do |user, password, url, table|

  table.hashes.each do |hash|
    http, request = get_http_request("POST",url,user,password)
    # support SKIP
    hash.each do |doc|
      if doc[1] == SKIP
        hash.delete(doc[0])
      end
    end
    request.set_form_data(hash)
    execute_http_request(http, request)
    begin
      @response.code.should == "201"
    rescue RSpec::Expectations::ExpectationNotMetError => error
      # better error msg with actual JSON vs just the diff from json_spec
      full_error = "#{error}\nHTTP response: #{@response.body}\n#{@response.header}"
      raise RSpec::Expectations::ExpectationNotMetError.new(full_error)
    end
  end
end


###### WHEN ##############

When /^"(.+):(.+)" sends (POST|PUT|PATCH) "(.+)" with "([^"]+)"$/ do |user,password,method,url,parameters|

  http, request = get_http_request(method,url, user, password)
  apply_form_parameters(request,parameters)
  execute_http_request(http,request)
end


When /^"(.+):(.+)" sends (OPTIONS|GET|DELETE|POST|PUT|PATCH) "([^"]+)"$/ do |user,password,method,url|
  http, request = get_http_request(method,url, user, password)
  execute_http_request(http,request)
end

When /^"(.+):(.+)" sends (POST|PUT|PATCH) "([^"]+)" with (CSV|text) file "([^"]+)"$/ do |user,password,method,url,content_type,file_name|
  http, request = get_http_request(method,url, user, password)
  case content_type
    when "CSV"
      request["content-type"] = "text/csv"
    when "text"
      request["content-type"] = "text/plain"
  end

  request.body = File.read(file_name)
  execute_http_request(http,request)
end

When /^"(.+):(.+)" sends (POST|PUT|PATCH) "([^"]+)" with (text|json|JSON|csv)$/ do |user,password,method,url,content_type,text|
  http, request = get_http_request(method,url, user, password)
  case content_type
    when "json","JSON"
      request["content-type"] = "application/json"
    when "text"
      request["content-type"] = "text/plain"
    when "csv"
      request["content-type"] = "text/csv"
  end
  request.body = text
  execute_http_request(http,request)
end


When /^I set HTTP header "([^"]*)" to "([^"]*)"$/ do |header_name, header_value|
  @http_headers[header_name] = header_value
end

When(/^"(.+):(.+)" sends (GET|POST) "([^"]+)" on admin port$/) do |user,password,method,url|
  # send it to all apps registered in the $config
  $config["http"].each do |doc|
    app_context = doc[0]
    http, request = get_http_request(method,url,user,password,true,app_context)
    execute_http_request(http, request)
  end
end

# executed against all admin ports in an app
When(/^"(.+):(.+)" sends POST "([^"]+)" with "([^"]+)" on admin port$/) do |user,password,url,parameters|
  $config["http"].each do |doc|
    app_context = doc[0]
    http, request = get_http_request("POST", url, user, password, true, app_context)
    apply_form_parameters(request,parameters)
    execute_http_request(http, request)
  end
end

# application-specific, executed on a single app admin port only
When(/^"(.+):(.+)" sends (GET|POST) "([^"]+)" on "(.+)" admin port$/) do |user,password,method,url,app_context|
  # send it to all apps registered in the $config 
  http, request = get_http_request(method,url,user,password,true,app_context)
  execute_http_request(http, request)
end

# application-specific, executed on a single app admin port only
When /^"(.+):(.+)" sends (POST|PUT|PATCH) "([^"]+)" with (text|JSON|form) on "(.+)" admin port$/ do |user,password,method,url,content_type,app_context,text|
  http, request = get_http_request(method,url, user, password,true,app_context)
  case content_type
    when "JSON"
      request["content-type"] = "application/json"
      request.body = text
    when "text"
      request["content-type"] = "text/plain"
      request.body = text
    when "form"
      request["content-type"] = "application/x-www-form-urlencoded"
      apply_form_parameters(request,text)
  end

  execute_http_request(http,request)
end


When(/^I sleep for (\d+)\s* seconds$/) do |seconds|
  sleep(seconds.to_i)
end

###### THEN #############

Then /^I expect HTTP code (\d+)\s*$/ do |code|
  begin
    @response.code.should == code
  rescue RSpec::Expectations::ExpectationNotMetError => error
    # better error msg with actual JSON vs just the diff from json_spec
    full_error = "#{error}\nHTTP response: #{@response.body}\n#{@response.header}"
    raise RSpec::Expectations::ExpectationNotMetError.new(full_error)
  end
end

Then /^I expect HTTP header "([^"]*)" equals "([^"]*)"$/ do |header_name, header_value|
  header_value = replace_memorized_variables(header_value)
  if header_name != SKIP and header_value != SKIP then
    @response[header_name].should == header_value
  end
end

Then(/^I expect JSON equivalent to$/) do |string|
  string = replace_memorized_variables(string,false)
  expected = JSON.parse(string)
  expected_field_names = get_unique_field_names(expected)

  actual = JSON.parse(last_json())
  actual_compare = remove_field_names(JSON.parse(last_json()), expected_field_names)

  # use json_spec comparison with diff built in
  begin
    JSON.dump(actual_compare).should be_json_eql(string)
  rescue RSpec::Expectations::ExpectationNotMetError => json_error
    # better error msg with actual JSON vs just the diff from json_spec
    full_error = "Actual full JSON:\n" + JSON.pretty_generate(actual) +
        "\n------------------------------\nActual JSON with expected fields only:\n" + JSON.pretty_generate(actual_compare) +
        "\n------------------------------\n" + json_error.to_s
    raise RSpec::Expectations::ExpectationNotMetError.new(full_error)
  end

end

Then(/^I expect JSON equal to$/) do |string|
  begin
    last_json.should be_json_eql(string)
  rescue RSpec::Expectations::ExpectationNotMetError => json_error
    # better error msg with actual JSON vs just the diff from json_spec
    full_error = "Actual JSON:\n" + JSON.pretty_generate(last_json) + "\n" + json_error.to_s
    raise RSpec::Expectations::ExpectationNotMetError.new(full_error)
  end
end


Then(/^I expect HTTP content contains "(.*)"$/) do |string|
  string = replace_memorized_variables(string)
  @response.body.should include(string)
end

Then(/^I expect HTTP content does not contain "(.*)"$/) do |string|
  if @response.body[string] != nil
    full_error = "Text '#{string}' found in response body:\n#{@response.body}"
    raise RSpec::Expectations::ExpectationNotMetError.new(full_error)
  end
end


Then(/^I expect HTTP content contains$/) do |string|
  string = replace_memorized_variables(string)
  begin
    @response.body.should include(string)
  rescue RSpec::Expectations::ExpectationNotMetError => json_error
    # better error msg with actual JSON vs just the diff from json_spec
    full_error = "Actual text:\n" + @response.body + "\n" + json_error.to_s
    raise RSpec::Expectations::ExpectationNotMetError.new(full_error)
  end
end

Then(/^I expect HTTP content matches/) do |string|
  begin
    /#{string}/.should match(@response.body)
  rescue RSpec::Expectations::ExpectationNotMetError => json_error
    # better error msg with actual JSON vs just the diff from json_spec
    full_error = "Actual text:\n" + @response.body + "\n" + json_error.to_s
    raise RSpec::Expectations::ExpectationNotMetError.new(full_error)
  end
end


# UTILITY METHODS

# loads the host and port from the app-specific context in the URL
def get_http(url, admin_port = false, app_context = nil)

  app_context = url.split("/")[1] if app_context == nil
  host = $config["http"][app_context]["host"]
  port = $config["http"][app_context]["port"]
  if admin_port
    port = $config["http"][app_context]["adminPort"]
  end

  http = Net::HTTP.new(host, port)

  # for large perf files
  http.continue_timeout = 30 * 60
  http.read_timeout = 30 * 60

  return http
end

# generates the appropriate HTTP request based on method
def get_http_request(method,url,user,password, admin_port = false, app_context = nil)
  http = get_http(url, admin_port, app_context)

  url = replace_memorized_variables(url)

  request = nil
  case method
    when "GET"
      request = Net::HTTP::Get.new(url)
    when "OPTIONS"
      request = Net::HTTP::Options.new(url)
    when "POST"
      request = Net::HTTP::Post.new(url)
    when "PUT"
      request = Net::HTTP::Put.new(url)
    when "PATCH"
      request = Net::HTTP::Patch.new(url)
    when "DELETE"
      request = Net::HTTP::Delete.new(url)
  end

  request.basic_auth(user,password)
  # default content type for all request, can be overriden
  request["content-type"] = "application/x-www-form-urlencoded"
  request["accept"] = "*/*"

  return http, request
end

# standard logic for all HTTP Request steps
def execute_http_request(http, request)
  @http_headers.each do |header|
    request[header[0]] = header[1]
  end

  @response = http.request(request)
  @http_headers = {}
end

def apply_form_parameters(request,parameters)
  # convert parameters to Ruby hash
  form = {}
  splitparams = parameters.split "&"
  splitparams.each do |part|
    keyValue = part.split "="
    key = keyValue[0].strip()
    value = keyValue[1]
    # support multiple parameters -> turn it into an array
    if !form.has_key?(key)
      form[key] = [value]
    else
      form[key] << value
    end
  end
  request.set_form_data(form)
end

# required by json_spec
def last_json
  @response.body
end

# find all the unique field names in a JSON document
def get_unique_field_names(json_doc, fields_set = SortedSet.new([]))

  json_doc.each{|k,v| fields_set.add(k)} if json_doc.is_a?(Hash)
  json_doc.each{|k,v| get_unique_field_names(json_doc[k],fields_set)} if json_doc.is_a?(Hash)
  json_doc.each {|doc| get_unique_field_names(doc, fields_set)} if json_doc.is_a?(Array)
  return fields_set
end

# removes the field names from the JSON doc that are not in the allowed list
def remove_field_names(json_doc, fields_set)

  json_doc.reject!{|k,v| not fields_set.include?(k)} if json_doc.is_a?(Hash)
  json_doc.each{|k,v| remove_field_names(json_doc[k],fields_set)} if json_doc.is_a?(Hash)
  json_doc.each {|doc| remove_field_names(doc, fields_set)} if json_doc.is_a?(Array)

  return json_doc
end

# takes a string and replaces all references to JsonSpec memory variables with actual values
def replace_memorized_variables(string_value, remove_quotes = true)
  # if JsonSpec variables are present, replace any potential references to them in the URL
  JsonSpec.memory.each do |doc|
    var_name = doc[0]
    # doc[1] is stored with quotes "....." so we have to remove them if told to (usually yes)
    var_value = doc[1]
    var_value = var_value.gsub('"',"") if remove_quotes
    string_value = string_value.gsub("%{#{var_name}}",var_value)
  end

  return string_value
end